Trustroots/trustroots
# Move css/less building to webpack, replace livereload with webpack hot reloading

I wanted to just rip out css/less and move to webpack, but then it lost live reloading, so I added it back in with webpack hot reloading, but that was tricky for the react components, but it works now! (just using very experimental library for now!).

## Changes

### Choosing which css/less files to load

#### Before

Various entries specified in `config/assets/default.js`, split into app and lib categories.

#### After

All via imports which are processed by webpack.

So in main.js there is:
```js
import './main.less';
```
Inside `main.less` it imports all the css files, plus less files from libraries (i.e. bootstrap), and also our global overrides.

Back in `main.js` it then loads all the `*.less` within `modules/`, using this fancy webpack context:
```js
importAll(require.context('../../../modules/', true, /\.less$/));
function importAll(r) {
  r.keys().forEach(r);
}
```
(this makes it similar to how we load the react components just below there)

### Other resources within css/less files

#### Before

They were all absolute paths which corresponds to the `public/` directory, e.g.:

```css
background-image: url(/img/avatar.png);
```

.. would be loaded by the browser from `<root>/public/img/avatar.png`

#### After

That (should) still works as before, _but_ as the files are imported via webpack it wants to load and process them too, and we use an alias to refer to `public/img` directory (configured in `config/webpack/webpack.config.js`. So, we might write:

```css
background-image: url(~img/avatar.png);
```

Now, webpack will pass that through it's loaders, and make it available at something like `public/assets/images/avatar-3a74587.png` (the hash is generated during the webpack processing, and the original css/less `url()` changed to match.

Also, it will inline the images using data-uri scheme if they are smaller than 10k (configurable, check webpack config again...). Not sure what the best value for that is.

So, everything in `public/assets` (served at `/assets`) is webpack output. In dev mode no files will actually be written, it's all served by webpack-dev-server. In production build, it will write files into `public/assets`.

### node.js server

#### Before

The trustroots server (the node.js express app running the backend) was running on port 3000, and you directly connect to it.

#### After

The node.js app is running on port 3001, and on 3000 is webpack dev server which does two things:
1. any requests to the js/css files it _would_ build, it will serve directly (from memory)
2. anything else it will proxy through to port 3001

### Other bits and pieces

#### No refresh on angular template change

Currently we no longer refresh on angular template changes (but will still do full page refreshes for other angular changes, controllers, etc), this will be fixed in the future when we load angular templates through webpack too.

#### Random fix to server test watching

Whilst removing livereload I fixed an issue! It had some code that ran when the server code changed, and it tried to check if it was a server test, and if so, only run that specific test. It didn't really work. It does now.

#### Bleeding edge react hot reloading plugin

The old way of hot loading react is going away (see [this thread](https://github.com/facebook/react/issues/16604)), but the new way didn't really get official webpack support yet. I tried using the old way, but it didn't really work with how we load the react components. So I tried the new way, it worked!

But the plugin to connect the new stuff with webpack (`@pmmmwh/react-refresh-webpack-plugin`) is so new, it has some rough edges, one of them makes incremental compilation _really_ slow, there is [an issue](https://github.com/pmmmwh/react-refresh-webpack-plugin/issues/20#issuecomment-565470955) for it, but for now you, if you want reasonable update speed, you have to comment out the `this.addDependency(path.resolve('./runtime/RefreshModuleRuntime'));` line in `node_modules/@pmmmwh/react-refresh-webpack-plugin/src/loader.js`. I could switch to one of it's forks (e.g. https://github.com/WebHotelier/webpack-fast-refresh), but I think this is a fast moving game... need to wait for it to settle, but maybe this hack is ok for now... it's still faster than the full refresh before probably and I think a fix will come soon.

## How to test

* `git checkout build-styles-in-webpack` (ok I think the work slightly expanded in scope since I picked the branch name!)
* `npm install`
* optional, comment out the `this.addDependency(path.resolve('./runtime/RefreshModuleRuntime'));` line in `node_modules/@pmmmwh/react-refresh-webpack-plugin/src/loader.js` (horrible, hack, see comment just above this for explanation)
* `npm start`
* see if everything seems ok!
* check hot reloading
  * visit http://localhost:3000/volunteering
  * open `./modules/pages/client/components/Volunteering.component.js`
  * edit something, e.g. `<h2>{t('Volunteering')}</h2>` -> `<h2>{t('Volunteering2')}</h2>`
  * see it update in your browser without a refresh!
  * if it doesn't work read the log output (wherever you ran `npm start`), and/or pop open a console in your browser and see if anything looks interesting
  * to try style reloading, visit http://localhost:3000/messages
  * open `./modules/messages/client/less/inbox.less`
  * fiddle around a bit... e.g. a nice `border: 10px solid red` on the `.threadlist`

If you use docker, the same should apply... (except editing the addDependency thing will be more painful.... as node_modules is only inside the container).

## TODO

- [x] integrate into main `npm start` scripts, so it JustWorks
- [x] remove unused less imports (a few I ended up importing higher up the chain... so can be removed from lower down)
- [x] ensure it spits assets out correctly during production builds (`publicPath` settings for style sometimes gets confusing)
- [x] check I removed now-unused libraries from package.json
- [x] check gulp is not watching files it doesn't need to be any more (now webpack does a lot of the watching)
- [x] ~document the need to install the react dev tools for the hot reload (and check first if that's actually true, I think it is...)~ it was a lie, you don't need react dev tools for it
- [x] make sure the docker setup works, re: ports, etc...
- [x] add back the webpack build for the service worker

## TODO in another PR

- [ ] load angular templates through webpack
- [ ] in order to support all the cool webpack splitting options and hashed filenames we need to tell the server which assets to load, I can steal/adapt what I did in foodsharing webpack config - it writes a json file with all the assets, then the server reads it, so it can generate the right js/css tags in the index.html (not needed right now if we stick to static filenames, so probably leave this out for now)
- [ ] we don't have many/any css/js optimization options set, maybe need some more of those...
